mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,1046 @@
+  forOp.walk([&](Operation *op) {
+    if (isa<AffineIfOp>(op))
+      iffound = 1;
+  });
+  return iffound;
+}
+
+/// <summary>
+/// Tests if a given AffineForOp loopnest has a rectangular shaped iteration
+/// space.
+/// </summary>
+/// <param name="loopnest"></param>
+/// <returns>Returns True if each AffineForOp in the loopnest has a constant
+/// upper and lower bound. Returns false otherwise.</returns>
+bool isRectangularAffineForLoopNest(
+    llvm::SmallVector<AffineForOp, 0> &loopnest) {
+  for (auto a : loopnest) {
+    if (!a.hasConstantUpperBound() || !a.hasConstantLowerBound())
+      return false;
+  }
+  return true;
+}
+
+/// <summary>
+/// Calculates Access Matrix for each memory access operation within the
+/// loopnest rooted at rootForOp by parsing the memory access expressions. For a
+/// memory access Ax+B, it calculates both A and B and populates the protected
+/// "loopAccessMatrices" and "constVector" properties of LoopInterchange object.
+/// </summary>
+/// <param name="rootForOp">Root of the loopnest</param>
+/// <param name="AffineForOpLoopNest"></param>
+/// <param name="loopVarIdMap">The map maintains an Id to each loop induction
+/// variable</param>
+void LoopInterchange::getAffineAccessMatrices(
+    AffineForOp rootForOp,
+    llvm::SmallVector<AffineForOp, 0> &AffineForOpLoopNest,
+    llvm::DenseMap<Value, int64_t> &loopVarIdMap) {
+  // for the mod / ceildiv / floordiv operations, consider them similar to
+  // access to dimensions.
+  this->loopAccessMatrices.clear();
+  this->constVector.clear(); // For Ax+B, this represents B
+  SmallVector<Operation *, 8> loadAndStoreOpInsts;
+  rootForOp.getOperation()->walk([&](Operation *opInst) {
+    if (isa<AffineLoadOp>(opInst) || isa<AffineStoreOp>(opInst)) {
+      loadAndStoreOpInsts.push_back(opInst);
+    }
+  });
+
+  unsigned numOps = loadAndStoreOpInsts.size();
+  for (unsigned i = 0; (i < numOps); ++i) {
+    auto *srcOpInst = loadAndStoreOpInsts[i];
+    MemRefAccess srcAccess(srcOpInst);
+    AffineMap map;
+    if (auto loadOp = dyn_cast<AffineLoadOp>(srcAccess.opInst))
+      map = loadOp.getAffineMap();
+    else if (auto storeOp = dyn_cast<AffineStoreOp>(srcAccess.opInst))
+      map = storeOp.getAffineMap();
+    SmallVector<Value, 8> operands(srcAccess.indices.begin(),
+                                   srcAccess.indices.end());
+
+    fullyComposeAffineMapAndOperands(&map, &operands);
+    map = simplifyAffineMap(map);
+    canonicalizeMapAndOperands(&map, &operands);
+
+    auto mapResults = map.getResults();
+
+    uint64_t noDims = map.getNumDims();
+    uint64_t noSymbols = map.getNumSymbols();
+
+    // #Rows in each accessMatrix: #Dimensions of the MemRef object
+    // #Columns in each accessMatrix: (noDims + noSymbols)
+
+    llvm::SmallVector<std::vector<int64_t>, 0> accessMatrix;
+    std::vector<int64_t> constVector(mapResults.size());
+
+    // accesssing the affine operations.
+    // all for a single instruction
```

Comment:
All comments have to be proper and grammatically correct sentences. Terminate with a full stop. Fix everywhere else.